File: core/trading/fee_model.py
"""
Realistic Fee Model Engine — Must be used by EVERY backtest and live trade.

Models the true round-trip cost of trading on Solana DEXs including:
    - AMM swap fees
    - AMM price impact (based on trade size vs pool liquidity)
    - MEV / sandwich attack exposure
    - Bags.fm perpetual creator fees (1% each direction)
    - Solana priority fees

Round-trip cost benchmarks:
    - HIGH tier (SOL/USDC):        0.3 – 0.6%
    - MID tier ($100K–$1M pool):   1 – 3%
    - MICRO tier ($10K–$100K):     3 – 8%
    - BAGS pre-graduation:         2% base + 2% creator fees + curve impact
    - BAGS post-graduation:        4 – 15% (thin new pool + 2% creator fees)
    - XSTOCK:                      1 – 3% (+ oracle spread off-hours)

A trade is only worth taking if:
    Expected_Edge > (total_round_trip_cost × 2.0)

Usage::

    from core.trading.fee_model import calculate_trade_cost, is_trade_viable
    from core.data.asset_registry import AssetClass

    cost = calculate_trade_cost(
        asset_class=AssetClass.NATIVE_SOLANA,
        pool_liquidity_usd=5_000_000,
        trade_size_usd=500,
    )
    print(f"Round-trip cost: {cost.total_round_trip_pct:.2%}")

    viable = is_trade_viable(expected_edge_pct=0.03, cost=cost)
"""

from __future__ import annotations

import logging
import os
from dataclasses import dataclass
from enum import Enum
from typing import Optional

logger = logging.getLogger(__name__)

# Import AssetClass — graceful fallback if asset_registry not available
try:
    from core.data.asset_registry import AssetClass
except ImportError:
    # Standalone usage without registry
    class AssetClass(Enum):  # type: ignore[no-redef]
        NATIVE_SOLANA = "native_solana"
        BAGS_BONDING_CURVE = "bags_bonding_curve"
        BAGS_GRADUATED = "bags_graduated"
        XSTOCK = "xstock"
        MEMECOIN = "memecoin"
        STABLECOIN = "stablecoin"


# ---------------------------------------------------------------------------
# Liquidity tiers
# ---------------------------------------------------------------------------

class LiquidityTier(Enum):
    """Classification by pool liquidity depth."""

    HIGH = "high"                   # Pool liquidity > $1M
    MID = "mid"                     # $100K – $1M
    MICRO = "micro"                 # $10K – $100K
    BAGS_PRE_GRAD = "bags_pre_grad"
    BAGS_POST_GRAD = "bags_post_grad"
    XSTOCK = "xstock"


def classify_liquidity(pool_liquidity_usd: float) -> LiquidityTier:
    """Classify a pool by its USD liquidity."""
    if pool_liquidity_usd >= 1_000_000:
        return LiquidityTier.HIGH
    if pool_liquidity_usd >= 100_000:
        return LiquidityTier.MID
    return LiquidityTier.MICRO


# ---------------------------------------------------------------------------
# Cost result
# ---------------------------------------------------------------------------

@dataclass
class TradeCost:
    """All-in cost breakdown for a single trade (one direction)."""

    dex_fee_pct: float              # AMM swap fee (e.g., 0.25% for Raydium)
    price_impact_pct: float         # AMM price impact for this trade size
    mev_cost_pct: float             # Expected MEV / sandwich cost
    creator_fee_pct: float          # Bags.fm creator fee (1% if applicable)
    priority_fee_sol: float         # Solana priority fee in SOL
    priority_fee_usd: float         # Priority fee converted to USD
    total_one_way_pct: float        # Sum of all percentage costs (one direction)
    total_round_trip_pct: float     # Both buy AND sell combined
    liquidity_tier: LiquidityTier

    def to_dict(self) -> dict:
        return {
            "dex_fee_pct": self.dex_fee_pct,
            "price_impact_pct": self.price_impact_pct,
            "mev_cost_pct": self.mev_cost_pct,
            "creator_fee_pct": self.creator_fee_pct,
            "priority_fee_sol": self.priority_fee_sol,
            "priority_fee_usd": self.priority_fee_usd,
            "total_one_way_pct": self.total_one_way_pct,
            "total_round_trip_pct": self.total_round_trip_pct,
            "liquidity_tier": self.liquidity_tier.value,
        }


# ---------------------------------------------------------------------------
# Fee parameters by tier
# ---------------------------------------------------------------------------

# DEX swap fees (Raydium/Orca standard)
DEX_FEE_BPS = {
    LiquidityTier.HIGH: 25,           # 0.25%
    LiquidityTier.MID: 25,
    LiquidityTier.MICRO: 30,          # Higher fee pools
    LiquidityTier.BAGS_PRE_GRAD: 100, # 1% bonding curve fee
    LiquidityTier.BAGS_POST_GRAD: 30,
    LiquidityTier.XSTOCK: 30,
}

# Expected MEV / sandwich cost by tier
MEV_COST_BPS = {
    LiquidityTier.HIGH: 5,            # 0.05% — large pools harder to sandwich
    LiquidityTier.MID: 30,            # 0.30%
    LiquidityTier.MICRO: 100,         # 1.0% — easy targets
    LiquidityTier.BAGS_PRE_GRAD: 50,  # 0.5%
    LiquidityTier.BAGS_POST_GRAD: 150,# 1.5% — graduation hype = MEV magnet
    LiquidityTier.XSTOCK: 20,         # 0.2%
}

# Default priority fee in SOL
DEFAULT_PRIORITY_FEE_SOL = 0.0005
DEFAULT_SOL_PRICE_USD = 150.0  # Fallback if live price unavailable


# ---------------------------------------------------------------------------
# Core calculation
# ---------------------------------------------------------------------------

def calculate_trade_cost(
    asset_class: AssetClass,
    pool_liquidity_usd: float,
    trade_size_usd: float,
    *,
    is_bags_token: bool = False,
    bags_creator_fee_pct: float = 0.0,
    priority_fee_sol: Optional[float] = None,
    sol_price_usd: Optional[float] = None,
    oracle_staleness_minutes: float = 0.0,
) -> TradeCost:
    """
    Calculate the realistic all-in cost for a trade.

    AMM price impact formula:
        impact = trade_size / (2 × pool_liquidity)

    Args:
        asset_class: Classification of the asset being traded
        pool_liquidity_usd: Total USD liquidity in the pool
        trade_size_usd: Notional USD value of the trade
        is_bags_token: Whether this is a Bags.fm token
        bags_creator_fee_pct: Per-direction creator fee (usually 1.0)
        priority_fee_sol: Override Solana priority fee
        sol_price_usd: Current SOL price for fee conversion
        oracle_staleness_minutes: For xStocks — spread widens with staleness
    """
    sol_price = sol_price_usd or DEFAULT_SOL_PRICE_USD
    prio_fee = priority_fee_sol if priority_fee_sol is not None else DEFAULT_PRIORITY_FEE_SOL

    # Determine liquidity tier
    if asset_class == AssetClass.BAGS_BONDING_CURVE:
        tier = LiquidityTier.BAGS_PRE_GRAD
    elif asset_class == AssetClass.BAGS_GRADUATED:
        tier = LiquidityTier.BAGS_POST_GRAD
    elif asset_class == AssetClass.XSTOCK:
        tier = LiquidityTier.XSTOCK
    else:
        tier = classify_liquidity(pool_liquidity_usd)

    # 1. DEX swap fee
    dex_fee_pct = DEX_FEE_BPS[tier] / 10_000

    # 2. Price impact: trade_size / (2 × pool_liquidity)
    if pool_liquidity_usd > 0:
        price_impact_pct = trade_size_usd / (2 * pool_liquidity_usd)
    else:
        price_impact_pct = 0.10  # 10% fallback for unknown liquidity

    # 3. MEV cost
    mev_cost_pct = MEV_COST_BPS[tier] / 10_000

    # 4. Creator fee (Bags.fm only)
    creator_fee_pct = 0.0
    if is_bags_token or asset_class in (AssetClass.BAGS_BONDING_CURVE, AssetClass.BAGS_GRADUATED):
        creator_fee_pct = bags_creator_fee_pct / 100 if bags_creator_fee_pct > 0 else 0.01  # Default 1%

    # 5. Oracle staleness spread (xStocks)
    staleness_spread = 0.0
    if asset_class == AssetClass.XSTOCK and oracle_staleness_minutes > 5:
        # Spread widens with staleness: 0.1% per 5 minutes of staleness
        staleness_spread = min(0.03, oracle_staleness_minutes * 0.0002)

    # 6. Priority fee as percentage of trade
    prio_fee_usd = prio_fee * sol_price
    prio_fee_pct = prio_fee_usd / trade_size_usd if trade_size_usd > 0 else 0.0

    # Total one-way cost
    total_one_way = (
        dex_fee_pct
        + price_impact_pct
        + mev_cost_pct
        + creator_fee_pct
        + staleness_spread
        + prio_fee_pct
    )

    # Round-trip = buy costs + sell costs
    # Most costs apply both directions; price impact doubles
    total_round_trip = (
        (dex_fee_pct * 2)
        + (price_impact_pct * 2)
        + (mev_cost_pct * 2)
        + (creator_fee_pct * 2)    # Creator fee on BOTH buy and sell
        + (staleness_spread * 2)
        + (prio_fee_pct * 2)       # Priority fee on both txns
    )

    return TradeCost(
        dex_fee_pct=dex_fee_pct,
        price_impact_pct=price_impact_pct,
        mev_cost_pct=mev_cost_pct,
        creator_fee_pct=creator_fee_pct,
        priority_fee_sol=prio_fee,
        priority_fee_usd=prio_fee_usd,
        total_one_way_pct=total_one_way,
        total_round_trip_pct=total_round_trip,
        liquidity_tier=tier,
    )


# ---------------------------------------------------------------------------
# Viability check
# ---------------------------------------------------------------------------

MINIMUM_EDGE_TO_COST_RATIO = 2.0


def is_trade_viable(
    expected_edge_pct: float,
    cost: TradeCost,
    min_ratio: float = MINIMUM_EDGE_TO_COST_RATIO,
) -> bool:
    """
    A trade is only worth taking if:
        Expected_Edge > (total_round_trip_cost × min_ratio)

    Default min_ratio is 2.0 — the edge must be at least 2x the cost.
    """
    if cost.total_round_trip_pct <= 0:
        return expected_edge_pct > 0
    ratio = expected_edge_pct / cost.total_round_trip_pct
    return ratio >= min_ratio


def edge_to_cost_ratio(expected_edge_pct: float, cost: TradeCost) -> float:
    """Return the ratio of expected edge to total round-trip cost."""
    if cost.total_round_trip_pct <= 0:
        return float("inf") if expected_edge_pct > 0 else 0.0
    return expected_edge_pct / cost.total_round_trip_pct


def minimum_edge_for_asset(
    asset_class: AssetClass,
    pool_liquidity_usd: float,
    trade_size_usd: float = 500.0,
    **kwargs,
) -> float:
    """
    Return the minimum expected edge (%) needed for a viable trade on this asset.

    This is: total_round_trip_cost × MINIMUM_EDGE_TO_COST_RATIO
    """
    cost = calculate_trade_cost(asset_class, pool_liquidity_usd, trade_size_usd, **kwargs)
    return cost.total_round_trip_pct * MINIMUM_EDGE_TO_COST_RATIO


# ---------------------------------------------------------------------------
# Backtest cost deduction helper
# ---------------------------------------------------------------------------

def deduct_costs(
    entry_price: float,
    exit_price: float,
    entry_cost: TradeCost,
    exit_cost: Optional[TradeCost] = None,
) -> float:
    """
    Apply realistic costs to a backtest trade and return net P&L percentage.

    If exit_cost is not provided, uses the same cost profile as entry.
    """
    if exit_cost is None:
        exit_cost = entry_cost

    gross_pnl_pct = (exit_price - entry_price) / entry_price if entry_price > 0 else 0.0

    # Deduct entry costs (buy side)
    net_entry = entry_price * (1 + entry_cost.total_one_way_pct)
    # Deduct exit costs (sell side)
    net_exit = exit_price * (1 - exit_cost.total_one_way_pct)

    net_pnl_pct = (net_exit - net_entry) / net_entry if net_entry > 0 else 0.0
    return net_pnl_pct

File: jarvis-sniper/src/lib/fee-model.ts
/**
 * Realistic Cost Model for Solana DEX Trading
 *
 * This module provides realistic fee and cost modeling for backtesting
 * and live trading. It addresses the critical issue that basic fee models
 * underestimate real trading costs on Solana AMMs.
 *
 * Key costs modeled:
 * - DEX swap fees (0.01-2% depending on pool tier)
 * - AMM price impact (trade_size / (2 × pool_liquidity))
 * - MEV/sandwich attack costs (0.5-3%+)
 * - Creator fees (Bags.fm 1% perpetual)
 * - Priority fees (network congestion)
 *
 * @module fee-model
 */

export type AssetLiquidityTier = 'high' | 'mid' | 'micro' | 'bags_pre_grad' | 'bags_post_grad' | 'xstock';

export interface TradeCost {
  /** AMM swap fee percentage */
  dexFeePct: number;
  /** AMM price impact percentage */
  priceImpactPct: number;
  /** Expected MEV/sandwich cost percentage */
  mevCostPct: number;
  /** Creator fee percentage (Bags.fm perpetual) */
  creatorFeePct: number;
  /** Priority fee in SOL */
  priorityFeeSol: number;
  /** Total cost percentage (sum of all percentage costs) */
  totalCostPct: number;
}

export interface FeeModelConfig {
  /** Asset liquidity tier */
  tier: AssetLiquidityTier;
  /** Pool liquidity in USD */
  poolLiquidityUsd: number;
  /** Trade size in USD */
  tradeSizeUsd: number;
  /** Is this a Bags.fm token */
  isBagsToken: boolean;
  /** Is token in pre-graduation bonding curve phase */
  isPreGraduation: boolean;
  /** Current SOL price for priority fee calculation */
  solPriceUsd: number;
  /** Market volatility level */
  volatility: 'low' | 'medium' | 'high' | 'extreme';
}

// Default DEX fees by pool tier
const DEX_FEE_BY_TIER: Record<AssetLiquidityTier, number> = {
  high: 0.003,      // 0.3% - SOL/USDC major pairs
  mid: 0.01,        // 1.0% - Established tokens
  micro: 0.02,       // 2.0% - Small caps
  bags_pre_grad: 0.003, // 0.3% - Bonding curve (lower fee during curve)
  bags_post_grad: 0.025, // 2.5% - Newly graduated (thin pool)
  xstock: 0.01,     // 1.0% - Tokenized stocks
};

// MEV cost estimates by volatility
const MEV_COST_BY_VOLATILITY: Record<string, number> = {
  low: 0.005,      // 0.5%
  medium: 0.01,    // 1.0%
  high: 0.02,      // 2.0%
  extreme: 0.03,   // 3.0%
};

/**
 * Calculate realistic trade costs for a given trade
 *
 * @param config - Fee model configuration
 * @returns TradeCost object with all cost components
 */
export function calculateTradeCost(config: FeeModelConfig): TradeCost {
  const {
    tier,
    poolLiquidityUsd,
    tradeSizeUsd,
    isBagsToken,
    isPreGraduation,
    solPriceUsd,
    volatility,
  } = config;

  // 1. DEX swap fee
  let dexFeePct = DEX_FEE_BY_TIER[tier];

  // Adjust for Bags token graduation status
  if (isBagsToken && !isPreGraduation) {
    dexFeePct = DEX_FEE_BY_TIER.bags_post_grad;
  } else if (isBagsToken && isPreGraduation) {
    dexFeePct = DEX_FEE_BY_TIER.bags_pre_grad;
  }

  // 2. AMM Price Impact
  // Formula: impact = trade_size / (2 × pool_liquidity)
  // This is the classic AMM price impact formula
  let priceImpactPct = 0;
  if (poolLiquidityUsd > 0) {
    priceImpactPct = tradeSizeUsd / (2 * poolLiquidityUsd);
    // Cap at reasonable maximum (50% for extremely illiquid)
    priceImpactPct = Math.min(priceImpactPct, 0.5);
  }

  // 3. MEV/Sandwich cost
  const mevCostPct = MEV_COST_BY_VOLATILITY[volatility];

  // 4. Creator fee (Bags.fm perpetual 1% each way = 2% round trip)
  let creatorFeePct = 0;
  if (isBagsToken) {
    creatorFeePct = 0.02; // 1% entry + 1% exit
  }

  // 5. Priority fee (dynamic based on volatility)
  const priorityFeeSol = calculatePriorityFee(volatility, solPriceUsd);

  // 6. Total cost
  const totalCostPct = dexFeePct + priceImpactPct + mevCostPct + creatorFeePct;

  return {
    dexFeePct,
    priceImpactPct,
    mevCostPct,
    creatorFeePct,
    priorityFeeSol,
    totalCostPct,
  };
}

/**
 * Calculate dynamic priority fee based on market conditions
 */
function calculatePriorityFee(
  volatility: 'low' | 'medium' | 'high' | 'extreme',
  solPriceUsd: number
): number {
  const baseFees: Record<string, number> = {
    low: 0.001,      // 0.001 SOL
    medium: 0.003,   // 0.003 SOL
    high: 0.005,     // 0.005 SOL
    extreme: 0.01,   // 0.01 SOL
  };

  const baseSol = baseFees[volatility];

  // Convert to USD for reporting
  return baseSol;
}

/**
 * Get round-trip cost benchmarks by tier
 * These are realistic estimates based on historical data
 */
export function getRoundTripCostBenchmark(tier: AssetLiquidityTier): {
  min: number;
  max: number;
  typical: number;
  description: string;
} {
  const benchmarks: Record<AssetLiquidityTier, { min: number; max: number; typical: number; description: string }> = {
    high: {
      min: 0.003,
      max: 0.006,
      typical: 0.0045,
      description: 'SOL/USDC, major pairs. Pool liquidity > $1M',
    },
    mid: {
      min: 0.01,
      max: 0.03,
      typical: 0.02,
      description: 'Established tokens. Pool liquidity $100K-$1M',
    },
    micro: {
      min: 0.03,
      max: 0.08,
      typical: 0.05,
      description: 'Small caps. Pool liquidity $10K-$100K',
    },
    bags_pre_grad: {
      min: 0.02,
      max: 0.04,
      typical: 0.03,
      description: 'Bonding curve pre-graduation. Includes creator fees',
    },
    bags_post_grad: {
      min: 0.04,
      max: 0.15,
      typical: 0.08,
      description: 'Just graduated, thin new pool + creator fees',
    },
    xstock: {
      min: 0.01,
      max: 0.05,
      typical: 0.02,
      description: 'Tokenized equities. Higher during off-hours',
    },
  };

  return benchmarks[tier];
}

/**
 * Check if a trade is worth taking given expected edge
 *
 * A trade should only be taken if:
 * Expected_Edge > (total_round_trip_cost × edge_to_cost_ratio)
 *
 * @param expectedEdgePct - Expected profit percentage
 * @param totalCostPct - Total round-trip cost percentage
 * @param minEdgeToCostRatio - Minimum edge-to-cost ratio (default 2.0)
 * @returns Whether the trade is worth taking
 */
export function isTradeWorthIt(
  expectedEdgePct: number,
  totalCostPct: number,
  minEdgeToCostRatio: number = 2.0
): boolean {
  // Must cover round-trip costs AND have minimum edge
  const breakEvenCost = totalCostPct * 2; // Entry + Exit
  const requiredEdge = breakEvenCost * minEdgeToCostRatio;

  return expectedEdgePct >= requiredEdge;
}

/**
 * Apply realistic costs to a backtest trade
 *
 * @param entryPrice - Entry price
 * @param exitPrice - Exit price
 * @param config - Fee model configuration
 * @returns Net profit after all costs
 */
export function calculateNetTradePnL(
  entryPrice: number,
  exitPrice: number,
  config: FeeModelConfig
): {
  grossPnlPct: number;
  netPnlPct: number;
  costs: TradeCost;
} {
  const costs = calculateTradeCost(config);

  // Gross P&L percentage
  const grossPnlPct = ((exitPrice - entryPrice) / entryPrice) * 100;

  // Net P&L after costs (entry cost + exit cost + price impact)
  const entryCost = costs.dexFeePct + (costs.isBagsToken ? costs.creatorFeePct / 2 : 0);
  const exitCost = costs.dexFeePct + costs.priceImpactPct + (costs.isBagsToken ? costs.creatorFeePct / 2 : 0);
  const totalCosts = entryCost + exitCost + costs.mevCostPct;

  const netPnlPct = grossPnlPct - (totalCosts * 100);

  return {
    grossPnlPct,
    netPnlPct,
    costs,
  };
}

/**
 * Determine liquidity tier based on pool data
 */
export function determineLiquidityTier(
  poolLiquidityUsd: number,
  isBagsToken: boolean,
  isPreGraduation: boolean
): AssetLiquidityTier {
  if (isBagsToken) {
    return isPreGraduation ? 'bags_pre_grad' : 'bags_post_grad';
  }

  if (poolLiquidityUsd > 1_000_000) return 'high';
  if (poolLiquidityUsd > 100_000) return 'mid';
  if (poolLiquidityUsd > 10_000) return 'micro';
  return 'micro'; // Default to micro for very small pools
}
